cmd/ctr/commands/images/crypt_utils.go

```diff
@@ -0,0 +1,183 @@
+	var (
+		gpgRecipients []string
+		pubkeys       []string
+		x509s         []string
+	)
+	for _, recipient := range recipients {
+		if strings.HasSuffix(recipient, ".pem") || strings.HasSuffix(recipient, ".der") {
+			tmp, err := ioutil.ReadFile(recipient)
+			if err != nil {
+				return nil, nil, nil, err
+			}
+			if isCertificate(tmp) {
+				x509s = append(x509s, base64.StdEncoding.EncodeToString(tmp))
+			} else {
+				pubkeys = append(pubkeys, base64.StdEncoding.EncodeToString(tmp))
+			}
+		} else {
+			gpgRecipients = append(gpgRecipients, recipient)
+		}
+	}
+	return gpgRecipients, pubkeys, x509s, nil
+}
+
+func processPrivateKeyFiles(keyFiles []string) ([]string, []string, error) {
+	var (
+		gpgSecretKeyRingFiles []string
+		privkeys              []string
+	)
+	// keys needed for decryption in case of adding a recipient
+	for _, keyfile := range keyFiles {
+		if strings.HasSuffix(keyfile, ".pem") || strings.HasSuffix(keyfile, ".der") {
+			tmp, err := ioutil.ReadFile(keyfile)
+			if err != nil {
+				return nil, nil, err
+			}
+			privkeys = append(privkeys, base64.StdEncoding.EncodeToString(tmp))
+		} else {
+			gpgSecretKeyRingFiles = append(gpgSecretKeyRingFiles, keyfile)
+		}
+	}
+	return gpgSecretKeyRingFiles, privkeys, nil
+}
+
+func createGPGClient(context *cli.Context) (encryption.GPGClient, error) {
+	gpgVersion := context.String("gpg-version")
+	v := new(encryption.GPGVersion)
+	switch gpgVersion {
+	case "v1":
+		*v = encryption.GPGv1
+	case "v2":
+		*v = encryption.GPGv2
+	default:
+		v = nil
+	}
+	return encryption.NewGPGClient(v, context.String("gpg-homedir"))
+}
+
+func getGPGPrivateKeys(context *cli.Context, gpgSecretKeyRingFiles []string, layerInfos []encryption.LayerInfo, mustFindKey bool, dcparameters map[string]string) error {
+	gpgClient, err := createGPGClient(context)
+	if err != nil {
+		return err
+	}
+
+	var gpgVault encryption.GPGVault
+	if len(gpgSecretKeyRingFiles) > 0 {
+		gpgVault = encryption.NewGPGVault()
+		err = gpgVault.AddSecretKeyRingFiles(gpgSecretKeyRingFiles)
+		if err != nil {
+			return err
+		}
+	}
+	return encryption.GPGGetPrivateKey(layerInfos, gpgClient, gpgVault, mustFindKey, dcparameters)
+}
+
+// cryptImage encrypts or decrypts an image with the given name and stores it either under the newName
+// or updates the existing one
+func cryptImage(client *containerd.Client, ctx gocontext.Context, name, newName string, cc *encryption.CryptoConfig, layers []int32, platformList []string, encrypt bool) (images.Image, error) {
+	s := client.ImageService()
+
+	image, err := s.Get(ctx, name)
+	if err != nil {
+		return images.Image{}, err
+	}
+
+	pl, err := platforms.ParseArray(platformList)
+	if err != nil {
+		return images.Image{}, err
+	}
+
+	lf := &encryption.LayerFilter{
+		Layers:    layers,
+		Platforms: pl,
+	}
+
+	newSpec, modified, err := images.CryptImage(ctx, client.ContentStore(), image.Target, cc, lf, encrypt)
```

Comment:
I can't seem to find the definition of this function but I would prefer not to have this be a bool. There should be discrete encrypt/decrypt functions.

Reply:
I changed this now so that `images` now exports `EncryptImage` and `DecryptImage` but then funnels right into the previous `cryptImage` function. The reason is that encryption and decryption share a lot of code. There are quite a few functions that start with `crypt` in the name but are not exported from `images`.